### DISCUSSION TITLE
Generate Change Language dialog from available CrowdIn languages

### DISCUSSION THREAD

**Discussion by okdistribute** [digidem/mapeo-desktop#595] [Ideas]
**Describe the idea**
Right now, the Change Language dialog is hard coded, only allowing Spanish, English, Portugese, and French. Instead, it should be generated from the available languages.

**What's your role? Where and how do you currently use the application?**
Tech @ Dd

**What is your ideal step by step workflow?**
The UX doesn't need to change, just the list of available languages should be larger to reflect all available languages.

**Have you thought about what it might look like?**
The code that is now hard coded is in `src/renderer/components/dialogs/ChangeLanguage.js`. Instead, this could be generated by scanning the disk for available languages in `messages/renderer/LANG.json` files. It's also possible this can also be detected from React Intl which is already imported into the file, which would be ideal, because it wouldn't involve any extra tooling.

**Comment by gmaclennan** [comment]
The code getting supported languages in Mapeo Mobile: https://github.com/digidem/mapeo-mobile/blob/develop/src/frontend/context/IntlContext.js
We show the list of languages with both the English name and local name in the settings screen: https://github.com/digidem/mapeo-mobile/blob/develop/src/frontend/screens/Settings/LanguageSettings.js
We use [`languages.json`](https://github.com/digidem/mapeo-mobile/blob/develop/src/frontend/languages.json) for the translated language names, however I have no recollection of where I found this file. Based on a brief search, similar files that I can find do not include regional variants (such as `af-NA` = "Afrikaans (Namibia)") which will be useful to support as Mapeo expands to new regions.
I think it's important to show the language list with local names as well as English: a non-English language speaker might not know their language name in English, and the English names are useful for someone who can not read local script setting up devices for users.
This is yet another piece of code that could be shared between Mapeo Mobile and Desktop, however Desktop has the addition of strings in the main process (the menus) which requires different code.
